### ISSUE TITLE
Explicit argument for `return` should be considered a best practice

### ISSUE THREAD

**Issue by stevecj** [koalaman/shellcheck#2886]
- [ X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

Maybe this is not a common misconception, but it bit me in the butt. I was piping output through the `tee_error` function (see below) in an `if` statement and not understanding why it was not executing the body of the `if` when the command at the head of the pipe was successful. I had assumed that `return` would behave the same as `return 0`.

Of course, the actual shell behavior is what makes the most sense, but this still tripped me up.

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

set -o pipefail

function tee_error
{
    IFS=$'\n' \read -rs line || return
    echo_and_log "${line}"
    while IFS=$'\n' \read -rs line; do
        echo_and_log "<e> ${line}" >&2
    done
}

# ...
```

#### Here's what shellcheck currently says:
(nothing)

#### Here's what I wanted or expected to see:
> Should supply an argument to `return`. By default, `return` returns the result of the preceding command and does not return 0 by default. To implement the default behavior explicitly, use `return $?`.

**Comment by ormaaj** [comment]
> I had assumed that return would behave the same as return 0.

I would say not a common misconception. No argument return is extremely common, and `return "$?"` is redundant and substituting that is really not best practice.

There exist some [weird situations](https://wiki.bash-hackers.org/dict/terms/exit_status?redirect=1#functions), of course, but the normal behavior of `return` transparently preserving the last status applies most of the time.

...are you aliasing `read`?? wat!

**Comment by stevecj** [comment]
Thanks for your reply, @ormaaj 

> ...are you aliasing `read`?? wat!

The script that is taken from is part of an installer to be run on systems that are not under our control, so the code is written in a paranoid manner, making as few assumptions as possible. We started implementing that kind of paranoia having been bitten by a customer exporting a function that overrides `whois`. Of course, I should have been using `id` instead of `whois`, and we changed that, but making that change does not fundamentally plug the hole.

**Comment by e-kwsm** [comment]
related (`$?`): #1995
